src/repo/remote.rs

```diff
@@ -0,0 +1,692 @@
+    let paths = fs::read_dir(base_path).map_err(Error::IOError)?;
+    paths
+        .filter_map(|e| e.ok())
+        .filter(|p| p.path().extension() == Some(OsStr::new(REMOTE_INDEX_EXTENSION)))
+        .map(|p| RemoteIndex::load(p.path()).reify(p.path().display()))
+        .collect()
+}
+
+/// Sends an HTTP GET request with the specified URL.
+///
+/// # Arguments
+/// * `url` - the URL to fetch
+/// * `timeout` - the timeout for the whole of the request
+/// * `size_limit` - the maximum size of the response body, before the
+///   connection is terminated
+/// * `if_modified_since` - Sets the value of the `If-Modified-Since` HTTP
+///   header
+fn open_remote_resource(
+    url: &Url,
+    timeout: Option<Duration>,
+    if_modified_since: Option<SystemTime>,
+) -> Result<Option<(usize, impl Read)>, Error> {
+    let mut agent_builder = AgentBuilder::new();
+    // Alternatively, we could have used Duration::MAX to indicate no timeout.
+    // Unfortunately, the ureq crashes at some unwrap() somewhere then MAX is
+    // provided.
+    if let Some(timeout) = timeout {
+        agent_builder = agent_builder.timeout(timeout);
+    }
+    let agent = agent_builder.build();
+
+    let mut request = agent.get(url.as_ref());
+    if let Some(if_modified_since) = if_modified_since {
+        request = request.set("If-Modified-Since", &format_http_date(if_modified_since));
+    }
+
+    let response = request.call().map_err(|e| Error::HttpError(e.into()))?;
+
+    let status = response.status();
+
+    let content_length = response
+        .header("Content-Length")
+        .unwrap_or("0")
+        .parse::<usize>()
+        .map_err(|_| {
+            Error::HttpError(
+                "The received Content-Length header contained an invalid value!".into(),
+            )
+        })?;
+
+    log::info!(
+        "HTTP GET {} -> {} (Content-Length: {})",
+        url,
+        status,
+        content_length
+    );
+
+    match status {
+        HTTP_STATUS_OK => Ok(Some((content_length, response.into_reader()))),
+        HTTP_STATUS_NOT_MODIFIED => Ok(None),
+        _ => Err(Error::HttpError(
+            format!(
+                "Response status code {} indicates failure!",
+                response.status()
+            )
+            .into(),
+        )),
+    }
+}
+
+fn read_remote_resource(
+    url: &Url,
+    timeout: Duration,
+    size_limit: u64,
+    if_modified_since: Option<SystemTime>,
+) -> Result<Option<Vec<u8>>, Error> {
+    if let Some(reader) = open_remote_resource(url, Some(timeout), if_modified_since)? {
+        let mut body = vec![];
+        reader
+            .1
+            // A malicious third-party could overwhelm might send a very
+            // large response.
+            .take(size_limit)
+            .read_to_end(&mut body)
+            .map_err(|e| Error::HttpError(e.into()))?;
+        Ok(Some(body))
+    } else {
+        Ok(None)
+    }
```

Comment:
Should this be rewritten as a map?

Reply:
Maybe, I'll see when I'm making the changes.